kotlin/workflow-ui-android/src/main/java/com/squareup/workflow/ui/ViewBackHandler.kt

```diff
@@ -0,0 +1,65 @@
+package com.squareup.workflow.ui
+
+import android.content.Context
+import android.content.ContextWrapper
+import android.view.View
+import android.view.View.OnAttachStateChangeListener
+import androidx.activity.ComponentActivity
+import androidx.activity.OnBackPressedCallback
+
+typealias BackPressHandler = () -> Unit
+
+var View.backPressedHandler: BackPressHandler?
+  get() {
+    val handler = getTag(R.id.view_back_handler)
+    check(handler is Function0<*>?) {
+      "Expected backPressedHandler of $this to be a Function0, found $handler."
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    return handler as? BackPressHandler
+  }
+  set(value) {
+    setTag(R.id.view_back_handler, value)
+    if (value == null) {
+      onBackPressedCallback?.let { removeOnAttachStateChangeListener(it) }
+      onBackPressedCallback = null
+    } else {
+      if (onBackPressedCallback == null) onBackPressedCallback =
+        AttachedOnBackPressedCallback(this).apply {
+          addOnAttachStateChangeListener(this)
+        }
+    }
+  }
+
+private var View.onBackPressedCallback: AttachedOnBackPressedCallback?
+  get() {
+    return getTag(R.id.view_on_back_pressed_callback) as? AttachedOnBackPressedCallback
+  }
+  set(value) {
+    onBackPressedCallback?.remove()
+    setTag(R.id.view_on_back_pressed_callback, value)
+  }
+
+private class AttachedOnBackPressedCallback(val view: View) : OnBackPressedCallback(true),
+    OnAttachStateChangeListener {
+  override fun handleOnBackPressed() {
+    view.backPressedHandler!!.invoke()
```

Comment:
Can't the handler function just also be passed to this class directly, so we don't need this null assertion?

Reply:
Ah, it doesn't because if the callback is already set we only bother to update the tag. Instead of that tricky layer of indirection, can we just store this class in the tag and make the handler function a mutable property on this class?